ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,777 @@
+    }
+  };
+
+  private static final Logger logger = LoggerFactory.getLogger(GetBlobOperation.class);
+
+  /**
+   * Construct a GetBlobOperation
+   * @param routerConfig the {@link RouterConfig} containing the configs for get operations.
+   * @param clusterMap the {@link ClusterMap} of the cluster
+   * @param responseHandler the {@link ResponseHandler} responsible for failure detection.
+   * @param blobIdStr the blob id associated with the operation in string form.
+   * @param futureResult the future that will contain the result of the operation.
+   * @param callback the callback that is to be called when the operation completes.
+   * @param operationCompleteCallback the {@link OperationCompleteCallback} to use to complete operations.
+   * @param time the Time instance to use.
+   * @throws RouterException if there is an error with any of the parameters, such as an invalid blob id.
+   */
+  GetBlobOperation(RouterConfig routerConfig, ClusterMap clusterMap, ResponseHandler responseHandler, String blobIdStr,
+      FutureResult<ReadableStreamChannel> futureResult, Callback<ReadableStreamChannel> callback,
+      OperationCompleteCallback operationCompleteCallback, Time time)
+      throws RouterException {
+    super(routerConfig, clusterMap, responseHandler, blobIdStr, futureResult, callback, time);
+    this.operationCompleteCallback = operationCompleteCallback;
+    this.blobIdFactory = new BlobIdFactory(clusterMap);
+    firstChunk = new FirstGetChunk(blobId);
+  }
+
+  /**
+   * {@inheritDoc}
+   * <br>
+   * If the callback for the operation was already called, then abort the reads from the channel and the caller
+   * will be notified as part of the read callback eventually. If not, the notification will happen as part of the
+   * operation completion callback.
+   */
+  @Override
+  void abort(Exception abortCause) {
+    if (operationCallbackInvoked.compareAndSet(false, true)) {
+      operationCompleteCallback.completeOperation(operationFuture, operationCallback, null, abortCause);
+    } else {
+      operationException.set(abortCause);
+      if (readCalled) {
+        completeRead();
+      }
+    }
+    operationCompleted = true;
+  }
+
+  /**
+   * Return the {@link MessageFormatFlags} to associate with a getBlob operation.
+   * @return {@link MessageFormatFlags#Blob}
+   */
+  @Override
+  MessageFormatFlags getOperationFlag() {
+    return MessageFormatFlags.Blob;
+  }
+
+  /**
+   * Do all that needs to be done (cleanup, notification, etc.) on chunk completion and mark the state of the chunk
+   * appropriately.
+   * @param chunk the chunk that has completed.
+   */
+  private void onChunkOperationComplete(GetChunk chunk) {
+    if (chunk.getChunkException() != null) {
+      // if operation callback was already called, then this exception will have to be notified as part of the
+      // read callback.
+      setOperationException(chunk.getChunkException());
+    }
+    if (chunk == firstChunk) {
```

Comment:
lets say we have 4 chunks in flight. and the 4th one resulted in exception. We would write all the remaining chunks to the asyncWritableChannel via maybeWriteToChannel() and then abort the operation since operationException was set due to 4th chunk